### PR TITLE
GRAMEX-128 ⁃ WIP: Statsmodels TSA

### PR DESCRIPTION
**Pull request by jaidevd**
```yaml
url:
  home:
    pattern: /$YAMLURL/autoreg
    handler: MLHandler
    kwargs:
      xsrf_cookies: false
      config_dir: $YAMLPATH
      data:
        url: $YAMLPATH/data/sunspots.csv
        index_col: YEAR
        parse_dates: [YEAR]
      model:
        dropna: false
        drop_duplicates: false
        class: AutoReg
        target_col: SUNACTIVITY
        params:
          lags: 5
```

Sample datasets are [here](https://teams.microsoft.com/_#/files/General?threadId=19%3AFcyuvPZLcvr5eJBaI5gB2FGQKpJze1txFjhQR6CvFN81%40thread.tacv2&ctx=channel&context=timeseries-datasets&rootfolder=%252Fsites%252FGramex-Platform%252FShared%2520Documents%252FGeneral%252Ftimeseries-datasets).

For all supported models, do the following:

* [x] See if the default training succeeds
* [x] Draw in-sample predictions
* [x] Retrain with different parameters
* [ ] Score predictions (RMSE only)
* [x] Try out-of-sample predictions / forecasting
* [x] Try multivariate timeseries

### Getting Predictions for endog-only models

```bash
# Timestamp format must match that in the training data
curl -X POST "http://localhost:9988/autoreg?_action=predict&start=1990-01-01&end=2021-10-31&index_col=time"
```

### Getting predictions for exog models:
```bash
curl -X POST -H 'Content-Type: application/json' -d @data.json \
    "http://localhost:9988/autoreg?_action=predict&index_col=time&start=1990-01-01&end=2021-10-31"
```

┆Issue is synchronized with this [Jira Bug](https://gramenertech.atlassian.net/browse/GRAMEX-128)

### CODE REVIEW

**Comment by jaidevd** [comment]
Available models:
* `AutoReg`
* `ARIMA`
* `SARIMAX`
* `ARDL`

File: gramex/sm_api.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
+import pandas as pd
+from statsmodels import api as sm
```
Comment (sanand0) [inline gramex/sm_api.py:2]: Statsmodels should be optional. Gramex should run without statsmodels as a dependency.

File: gramex/handlers/mlhandler.py [hunk 2]
```diff
@@ -367,7 +385,16 @@ def _train(self, data=None):
...
+        if isinstance(self.model, BaseModel):
+            index = self.get_argument('index_col')
+            if data.index.name != index and index in data:
+                data[index] = pd.to_datetime(data[index])
+                data.set_index(index, verify_integrity=True, inplace=True)
+            result = _fit(
+                self.model, None, data, self.model_path, target_col=target_col
+            ).res.summary().as_html()
+            self.set_header('Content-Type', 'text/html')
+        elif not isinstance(self.model[-1], TransformerMixin):
```
Comment (sanand0) [inline gramex/handlers/mlhandler.py:397]: ?_params should report the statsmodels parameters.

File: gramex/handlers/mlhandler.py [hunk 3]
```diff
@@ -271,6 +276,19 @@ def _transform(self, data, **kwargs):
...
+                start, end = self.get_argument('start'), self.get_argument('end')
+                index = self.get_argument('index_col')
+                if data.index.name != index and index in data:
+                    data[index] = pd.to_datetime(data[index])
+                    data.set_index(index, verify_integrity=True, inplace=True)
+                tcol = self.get_opt('target_col')
+                if tcol in data:
+                    data.pop(tcol)
+                return self.model.predict(start, end, exog=data)
+            return self.model.predict(**data.iloc[0].to_dict())
```
Comment (sanand0) [inline gramex/handlers/mlhandler.py:290]: Move this code into sm_api.py

File: gramex/handlers/mlhandler.py [hunk 4]
```diff
@@ -367,7 +385,16 @@ def _train(self, data=None):
...
         self.model = self._assemble_pipeline(data, force=True)
-        if not isinstance(self.model[-1], TransformerMixin):
+        if isinstance(self.model, BaseModel):
+            index = self.get_argument('index_col')
+            if data.index.name != index and index in data:
+                data[index] = pd.to_datetime(data[index])
+                data.set_index(index, verify_integrity=True, inplace=True)
+            result = _fit(
+                self.model, None, data, self.model_path, target_col=target_col
+            ).res.summary().as_html()
```
Comment (sanand0) [inline gramex/handlers/mlhandler.py:395]: Let's move this too into sm_api.py

File: gramex/handlers/mlhandler.py [hunk 5]
```diff
@@ -271,6 +276,19 @@ def _transform(self, data, **kwargs):
...
+                index = self.get_argument('index_col')
+                if data.index.name != index and index in data:
+                    data[index] = pd.to_datetime(data[index])
+                    data.set_index(index, verify_integrity=True, inplace=True)
+                tcol = self.get_opt('target_col')
+                if tcol in data:
+                    data.pop(tcol)
+                return self.model.predict(start, end, exog=data)
+            return self.model.predict(**data.iloc[0].to_dict())
+        app_log.critical(type(self.model))
```
Comment (jaidevd) [inline gramex/handlers/mlhandler.py:291]: Remove this.

File: gramex/handlers/mlhandler.py (hunk 2 above)
Comment (jaidevd) [inline gramex/handlers/mlhandler.py:397]: There has to be a more elegant way of checking if the estimator at the end of a pipeline, if it is a pipeline, is a transformer or not, _without_ needing the model object to be subscriptable. If nothing else, `mclass` exists in config.json or gramex.yaml.

**Comment by jaidevd** [comment]
The new spec should look like this:

```
url:
  home:
    pattern: /$YAMLURL/
    handler: MLHandler
    kwargs:
      config_dir: $YAMLPATH
      data:
        url: $YAMLPATH/data/sunspots.csv
        index_col: YEAR  # Infer the `parse_dates` argument directly from this!
      model:
        dropna: false
        drop_duplicates: false
        class: SARIMAX   # Only value allowed, anything else gets interpreted as sklearn
        target_col: SUNACTIVITY
        stl:
          components: ['season', 'trend']   # Can be any two of {'season', 'trend', 'residue'}. If ignored, or if all three are provided, don't do STL.
          kwargs: # other STL kwargs
```

**Comment by jaidevd** [comment]
Closing in favour of #507 

Some useful work on the templates is still here, will come back to it.

**Pull request closed without merging**